src/assign.c

```diff
@@ -152,13 +152,21 @@ static SEXP shallow(SEXP dt, SEXP cols, R_len_t n)
   R_len_t i,l;
   int protecti=0;
   SEXP newdt = PROTECT(allocVector(VECSXP, n)); protecti++;   // to do, use growVector here?
-  //copyMostAttrib(dt, newdt);   // including class
-  DUPLICATE_ATTRIB(newdt, dt);
+  SHALLOW_DUPLICATE_ATTRIB(newdt, dt);
```

Comment:
`SHALLOW_DUPLICATE_ATTRIB` may not be available in R 3.1.0, this is the file where that macro is defined in current R version, it seems to be missing in 3.1 https://github.com/wch/r-source/blob/R-3-1-branch/src/main/coerce.c

Reply:
Looks like you're right, it doesn't show up in Rinternals.h for the 3.1 branch. It is only available starting from the 3.3 branch. Do you think lifting the code from https://github.com/wch/r-source/blob/bf4cf8912596162a3ae5a2b7da1a74220105b172/src/main/memory.c (I believe this is the actual definition?) would be an option? I had tested with `copyMostAttrib` which seems to be functionally the same but of course it doesn't check the object and S4 bits and leaves out the names. But at least here it would work I think. I was just a little worried there and opted for `SHALLOW_DUPLICATE_ATTRIB` because that the closest to how it was before.

Reply:
```c
SET_ATTRIB(newdt, shallow_duplicate(ATTRIB(dt)));
```
could possibly work. `ATTRIB` gives `LISTSXP` so will play well with `shallow_duplicate`.

Reply:
You may find this example useful
https://github.com/Rdatatable/data.table/pull/4370/files#diff-293a9e4361f16756700c7a5baddbabafR19-R34

Reply:
Yes, that is essentially how `SHALLOW_DUPLICATE_ATTRIB` is defined so it ought to work (minus the ATTRIB on newdt) and it's what you do in your link. Thanks for that btw. So I'll change that line.